### PR TITLE
Upgrade the PyStub code to handle synthetic GeneratorParams

### PR DESCRIPTION
**Pull request by steven-johnson**
PyStub is the code that lets Python code call a (C++) Generator; this upgrades the code to allow passing in of 'synthetic' GeneratorParams (e.g. buffer.type), which it previously didn't handle.

At the moment this doesn't make a big difference (I'm not aware of ~any meaningful amount of code outside of the Python tests that use PyStub), but this will streamline some possible changes in an experimental branch and make future diffs less painful.

### CODE REVIEW

File: python_bindings/correctness/pystub.py
Comment (alexreinking) [inline python_bindings/correctness/pystub.py:185]: Why is this the case? Can't it be something like `untyped_buffer_output={"type": "uint8"}`?

Comment (steven-johnson) [reply python_bindings/correctness/pystub.py:185]: eh... I guess it could be? I avoided that mainly because when I wrote this I thought I'd want to implement the same thing for C++ stubs, in which case having a common syntax would be desirable, and the __ approach is more straightforward to manage for C++. I'll give your suggestion a try and see how it feels.

Comment (steven-johnson) [reply python_bindings/correctness/pystub.py:185]: Actually, there *is* a very good reason that syntax won't work as-is: inputs that also need their type (etc) set. Currently:

    untyped_buffer_input=constant_image,  // line 173 in this PR
    untyped_buffer_input__type="uint8",  // line 186 in this PR

With this proposal, we'd have the obviously-illegal:

    untyped_buffer_input=constant_image, 
    untyped_buffer_input={"type": "uint8"},

We could finesse this in various ways, of course, e.g.

    # require a dict that includes value as well as type, dim, etc for these cases
    untyped_buffer_input={"value": constant_image, "type": "uint8"},

...but that feels weird and awkward IMHO.

Comment (alexreinking) [reply python_bindings/correctness/pystub.py:185]: I see your point and I keep going back and forth on whether using `__` or a dictionary like this is _more_ weird and awkward. Double underscores generally indicate language-level magic in Python, which isn't really the case here. On the other hand, the `"value"` key isn't really mirrored anywhere else in the API.

Why can't we get the type out of the `constant_image` object? When it's created on line 150, it's typed with `hl.UInt(8)`.

Comment (steven-johnson) [reply python_bindings/correctness/pystub.py:185]: > Why can't we get the type out of the constant_image object?

(sorta-topic-change to other work going on elsewhere...)

We *can*, the question is whether we *should*: there is a distinction between "is passed-in-argument well-typed" (yes, it is), vs "is the type required by the receiver well-defined or not" (in this case, not).

Currently, in C++ Generators (and by extension, the almost-never-used C++-based stubs), we take the approach that we don't require the receiver to be explicitly typed iff we can infer a concrete type from the passed-in-argument. This seems appealing, and works, but has several drawbacks that I'd like to avoid in the still-work-in-progress Python generator design:
- It complicates the internal code for Generators, since we may no longer know the type constraints on Inputs/Outputs until you actually execute the Generator to produce the IR.
- It impairs the readability for Generators, since you may not be able to look at the code (or even the code + the build file containing GeneratorParam values) and know what the expected types are. (For 'types', also include 'dimensions' and 'array-size', of course.)

IMHO a superior design would have been to require that all 'untyped' things be explicitly specified via GeneratorParams, so that no Generator code must be executed to deduce the signature of the resulting function. Trying to change that design point for C++ Generators would be a fool's errand at this point, but I'd really like to make this a design point to enforce in Python.

Comment (steven-johnson) [reply python_bindings/correctness/pystub.py:185]: > I see your point and I keep going back and forth on whether using `__` or a dictionary like this is _more_ weird and awkward. Double underscores generally indicate language-level magic in Python, which isn't really the case here. 

Yeah, I see your point and don't disagree. 

Arguably, the 'bug' here is that 'synthetic' params like this were created with a '.' in their names as a way to ensure they didn't overlap with any other param, since '.' was not and is not a valid user-provided name -- since there wasn't (and still isn't) a way to specify these for C++ stubs (which are rarely used, a nightmare to implement and use, and probably will never be enhanced), this isn't a limitation, but for Python-based stubs (which are potentially *much* more useful, for various reasons), being able to support these would be quite useful.

So the real question here is: is there a syntax that fits reasonably with Python named-argument convention which is reasonable?

Comment (steven-johnson) [reply python_bindings/correctness/pystub.py:185]: Any further thoughts on this? I'm not especially happy with the `__` approach, but lack a better idea at the moment...

**Review by alexreinking — APPROVED** [review]
I don't love the `__`, but now that we're on the other side of the Halide 14 window, I feel like we're safe to change this if inspiration strikes and we find a better solution. Better not to let this go stale.

**Comment by steven-johnson** [comment]
> Better not to let this go stale

This will keep for a while. I'd rather get it right before landing it, and some of the related design is still in flux, so I'm gonna hold off until there's something we actually like, or we have a pressing need for it.

**Comment by alexreinking** [comment]
Sounds good. I re-requested a review from myself so it doesn't get merged accidentally.

**Comment by steven-johnson** [comment]
After further thought, I think this is actually moot for this use case -- when calling the Stub, we can always infer the type(s) (etc) that we need from the actual inputs. Closing.

**Pull request closed without merging**
